src/dualnumber.jl

```diff
@@ -624,55 +199,55 @@ end
 
 for pred in BINARY_PREDICATES
     @eval begin
-        @define_binary_dual_op(
-            Base.$(pred),
-            $(pred)(value(x), value(y)),
-            $(pred)(value(x), y),
-            $(pred)(x, value(y))
-        )
+        Base.$(pred)(x::Dual, y::Dual) = $pred(value(x), value(y))
+        Base.$(pred)(x::Number, y::Dual) = $pred(value(x), value(y))
+        Base.$(pred)(x::Dual, y::Number) = $pred(value(x), value(y))
     end
 end
 
 ########################
 # Promotion/Conversion #
 ########################
 
-Base.@pure function Base.promote_rule(::Type{Dual{T1,V1,N1}},
-                                      ::Type{Dual{T2,V2,N2}}) where {T1,V1,N1,T2,V2,N2}
+#T1 < T2  if T1 = Tag{Tag{Nothing}} but T2 = Tag{Tag{Tag{Nothing}}}
+
+Base.@pure function Base.promote_rule(::Type{Dual{T1,V1,P1}},
```

Comment:
Are you sure we need this `@pure`.
This is a illegal usage as it calls generic functions.
And its not a particularly a safe usage illegal usage at that.
In that if someone wants to add support for `V1` or `V2`, being their fancy new numeric type (say `ArbFloat`).
by overloading `promote_type(::V, ::Dual)` then bad things will happen.

Reply:
You are right, we shouldn't use `@pure` here. This is from ForwardDiff.jl, BTW.